## 1. Introduction

For a new self-driving car, we need an implementation of a high-precision pi: ChatGPT v4 suggests the following implementation for computing pi in Python, including a unit test. The code is packed in the two files `estimate_pi.py` and `test_estimate_pi.py`. 

Run the existing test using your shell (every cell starting with an `!` will be executed in your OS's shell). 

In [6]:
!python estimate_pi.py 
!python test_estimate_pi.py

3.14238


.
----------------------------------------------------------------------
Ran 1 test in 0.992s

OK


What is problematic with that test ChatGPT created for us and would you address this problem? (**1 point**)

    The problem is the function used in row 8 in test_estimate_pi.py; "self.assertAlmostEqual(pi_expected, pi_actual, delta=0.01).


    Problem.
    The test will always return OK as suggested by the documentation of the function. that is because of the asserAlmostEqual() in row 8 in the test. it will say that the test is "OK" if the difference is less than delta=0.01. meaning that the error have to be very large for the difference to be MORE then 0.01.
    
    Solution.
    1. use a lower delta. to get a more accurate estimation
    2. experiment with other Tests

## 2. Coverage

### 2.1 Statement coverage
Compute the statement coverage of the program using [`coverage.py`](https://coverage.readthedocs.io/en/latest/index.html). (**2 points**)

In [7]:
!coverage run test_estimate_pi.py
!coverage report -m

.
----------------------------------------------------------------------
Ran 1 test in 2.755s

OK


Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py       9      0   100%
---------------------------------------------------
TOTAL                    35     12    66%


How can we interprete the results?

    Interpretation
    We can see how much of the main program the test really tests. Here we see what rows are being run when the functions are being called and the rows that are being missed from the file. Coverage shows us the percentage of rows that are being run by the test.

    100% coverage of the test, meaning that the single test that we have, does run but that single test does only cover 54% of the file it tests. meaning that there are much more test that can be done. we also can see the rows that have not been tested, in this case rows: 18-19, 23-36
    

### 2.2 Branch coverage
Now compute the branch coverage of the program using [`coverage.py`](https://coverage.readthedocs.io/en/latest/index.html). (**2 points**)

In [8]:
!coverage run --branch test_estimate_pi.py
!coverage report -m

.
----------------------------------------------------------------------
Ran 1 test in 3.278s

OK


Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py       9      0      2      1    91%   46->exit
-----------------------------------------------------------------
TOTAL                    35     12     12      2    62%


How can we interprete the results?

    We could interprete branch coverage as how much of the logical paths that has been covered, not just how many lines, but also how it is covered in each statements

    91% coverage of the test, meaning that the single test that we have, does run but that single test does only cover 53% of the file it tests. meaning that there are much more test that can be done similar to statement coverage. we also can see the rows that have not been tested, in this case rows: 18-19, 23-36 in estimate_pi.py and 46->exit test_estimate_pi.py

## 3. More unit tests

Add two more unit tests with the principles you learned in the lecture. Describe what principle you have used. (**3 points**)


    Test1: Confusion test
        Self validating, because of assertion. we check so that the program does not produce another constant as Eulers number. 
    
    Test2: type test
        Self validating, because of assertion

## 4. Mocking

We want to store the resulting number persistently on our file system. We use the following class. 

In [4]:
class PiFileWriter:
    @staticmethod
    def write(content, file_path):
        with open(file_path, 'w') as file:
            file.write(content)

Implement a test double for `PiFileWriter` and add your implementation to `test_estimate_pi.py`. Discuss what type of test double you have implemented. (**3 points**)

    we have implemented a test double of type "FAKE". we create a class that represent a .txt file with Variables written_content and written_path. we created a static method that writes and stores the informaton in the variables. Because a test double are supposed to be independent, we have chosen to hard code the pi_value. 

Name three other types of unit tests you would want to mock and explain why. (**1 point**)

    # 1
    * Logger and Sensor. Sensor takes time to return a value, so we MOCK the sensor that has pre recorded values that are similar to the sensor. This will give us a return instantly for our test.

    # 2
    * Register users on a webpage. We want to test if the username registration works correctly, instead of running the webserver all the time, or request information all the time, we can MOCK the usernames, so that the MOCK sends us randomized or hardcorded usernames. This will save resources and time. Also we can MOCK confirmation emails.

    # 3
    * Borrowing school books. You MOCK the system that keeps track of how long you are able to loan a book. The MOCK in this case will return a specified (hardcoded) date. 

## 5. Coverage revisited

Rerun statement and branch coverage and discuss the differences and changes. (**3 points**)

In [ ]:
!coverage run test_estimate_pi.py
!coverage report -m


!coverage run --branch test_estimate_pi.py
!coverage report -m

....
----------------------------------------------------------------------
Ran 4 tests in 7.794s

OK


Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py      30      2    93%   26-27
---------------------------------------------------
TOTAL                    56     14    75%


....
----------------------------------------------------------------------
Ran 4 tests in 8.123s

OK


Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py      30      2      2      1    91%   26-27, 46->exit
-----------------------------------------------------------------
TOTAL                    56     14     12      2    71%


statment Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py       9      0   100%
---------------------------------------------------
TOTAL                    35     12    66%


branchName                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py       9      0      2      1    91%   46->exit
-----------------------------------------------------------------
TOTAL                    35     12     12      2    62%


--------------------------------------------------------------

OK
stmName                  Stmts   Miss  Cover   Missing
---------------------------------------------------
estimate_pi.py           26     12    54%   18-19, 23-36
test_estimate_pi.py      30      2    93%   26-27
---------------------------------------------------
TOTAL                    56     14    75%
....
----------------------------------------------------------------------
Ran 4 tests in 8.123s

OK
branName                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
estimate_pi.py           26     12     10      1    53%   18-19, 23-36
test_estimate_pi.py      30      2      2      1    91%   26-27, 46->exit
-----------------------------------------------------------------
TOTAL                    56     14     12      2    71%





    Why we wont get 100% covarage in estimate_pi is becasue we created a test double in test_estimate_file. so it will never run the Original


    Statement Coverage rose from 66% to 75% and Branch Coverage rose from 62% to 71%. We added more tests in the test file which leads to more coverage in total. the class PiFileWriter was never tested when we ran the coverage code in question 2. 

    the coverage in the file estimate_pi are unchanged to 54% statement and 53% branch coverage because we only tested things that were already in the file. 
    We created a test dubble which inherit the main function of the class PiFileWriter. 



## 6. Submit to Canvas

Almost done, but the most tricky part is missing: submitting. :)

Before submitting, make sure
- you don't use external libraries except `coverage.py`
- the notebook runs straight through
- your test code works
- your code is readable and follows the Python coding conventions

- Submit these files to canvas:
    - `assignment_1.ipynb`
    - `test_estimate_pi.py`
